test/framework/control_plane.go

```diff
@@ -34,213 +34,276 @@ import (
+}
 
+// WaitForClusterDeletedInput is the input for WaitForClusterDeleted.
+type WaitForClusterDeletedInput struct {
+	Getter  Getter
+	Cluster *clusterv1.Cluster
+}
+
+// WaitForClusterDeleted waits until the cluster object has been deleted.
+func WaitForClusterDeleted(ctx context.Context, input WaitForClusterDeletedInput, intervals ...interface{}) {
+	By(fmt.Sprintf("waiting for cluster %s to be deleted", input.Cluster.GetName()))
 	Eventually(func() bool {
-		clusters := clusterv1.ClusterList{}
-		if err := mgmtClient.List(ctx, &clusters); err != nil {
-			fmt.Println(err.Error())
-			return false
+		cluster := &clusterv1.Cluster{}
+		key := client.ObjectKey{
+			Namespace: input.Cluster.GetNamespace(),
+			Name:      input.Cluster.GetName(),
 		}
-		return len(clusters.Items) == 0
-	}, input.DeleteTimeout, eventuallyInterval).Should(BeTrue())
+		return apierrors.IsNotFound(input.Getter.Get(ctx, key, cluster))
+	}, intervals...).Should(BeTrue())
+}
+
+// AssertAllClusterAPIResourcesAreGoneInput is the input for AssertAllClusterAPIResourcesAreGone.
+type AssertAllClusterAPIResourcesAreGoneInput struct {
+	Lister  Lister
+	Cluster *clusterv1.Cluster
+}
 
+// AssertAllClusterAPIResourcesAreGone ensures that all known Cluster API resources have been remvoed.
+func AssertAllClusterAPIResourcesAreGone(ctx context.Context, input AssertAllClusterAPIResourcesAreGoneInput) {
 	lbl, err := labels.Parse(fmt.Sprintf("%s=%s", clusterv1.ClusterLabelName, input.Cluster.GetClusterName()))
 	Expect(err).ToNot(HaveOccurred())
-	listOpts := &client.ListOptions{LabelSelector: lbl}
+	opt := &client.ListOptions{LabelSelector: lbl}
 
 	By("ensuring all CAPI artifacts have been deleted")
-	ensureArtifactsDeleted(ctx, mgmtClient, listOpts)
-}
 
-func ensureArtifactsDeleted(ctx context.Context, mgmtClient client.Client, opt client.ListOption) {
-	// assertions
 	ml := &clusterv1.MachineList{}
-	Expect(mgmtClient.List(ctx, ml, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, ml, opt)).To(Succeed())
 	Expect(ml.Items).To(HaveLen(0))
 
 	msl := &clusterv1.MachineSetList{}
-	Expect(mgmtClient.List(ctx, msl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, msl, opt)).To(Succeed())
 	Expect(msl.Items).To(HaveLen(0))
 
 	mdl := &clusterv1.MachineDeploymentList{}
-	Expect(mgmtClient.List(ctx, mdl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, mdl, opt)).To(Succeed())
 	Expect(mdl.Items).To(HaveLen(0))
 
 	kcpl := &controlplanev1.KubeadmControlPlaneList{}
-	Expect(mgmtClient.List(ctx, kcpl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, kcpl, opt)).To(Succeed())
 	Expect(kcpl.Items).To(HaveLen(0))
 
 	kcl := &cabpkv1.KubeadmConfigList{}
-	Expect(mgmtClient.List(ctx, kcl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, kcl, opt)).To(Succeed())
 	Expect(kcl.Items).To(HaveLen(0))
 
 	sl := &corev1.SecretList{}
-	Expect(mgmtClient.List(ctx, sl, opt)).To(Succeed())
+	Expect(input.Lister.List(ctx, sl, opt)).To(Succeed())
 	Expect(sl.Items).To(HaveLen(0))
 }
```

Comment:
Might be good eventually to update this to use a dynamic client and discovery to check for any resources with cluster-api labels.  That would allow for providers to potentially consume without having to handle their own resources and also allow for future use of base k8s resources (such as configmap).